### ISSUE TITLE
Handling a custom mimetype in IRkernel

### ISSUE THREAD

**Issue by jlorince** [IRkernel/IRkernel#701]
Hello!

We have custom dataframe display code we're using in a python notebook, and we're working on porting the functionality to R using IRkernel, but having some difficulties. In python, we have the following:

```
from IPython.core.formatters import BaseFormatter
from traitlets import Unicode, ObjectName

DISPLAY_TABLE_MIMETYPE = "our_custom_mimetype"

class DisplayTable(self, df, ...):
    # implements the actual DisplayTable object, which is essentially just a dataframe with some extra properties

class DisplayTableJSONFormatter(BaseFormatter):
    format_type = Unicode(DISPLAY_TABLE_MIMETYPE)
    print_method = ObjectName("_repr_display_table_")

def display_table_json(display_table):
    # some custom logic for converting a DisplayTable object to JSON
    return json.dumps(...)

get_ipython().display_formatter.formatters[DISPLAY_TABLE_MIMETYPE] = DisplayTableJSONFormatter(parent=get_ipython().display_formatter)
get_ipython().display_formatter.formatters[DISPLAY_TABLE_MIMETYPE].for_type(DisplayTable, display_table_json)
```

In Python this all works fine; that is, we can have a cell with the contents `DisplayTable(df, ...)` and it renders as expected (we have custom logic in the UI that knows how to parse and render the JSON, assuming the payload has our custom mimetype).

The problem is, I'm not sure how to achieve the same effect in R. I can use reticulate to import the `DisplayTable` class as described above, but what's the correct way to register things with the IRkernel such that objects of the type `DisplayTable` will result in JSON with our custom mimetype?

Sorry if this question doesn't make tons of sense....still trying to wrap my head around all this...

- [x] I followed the [official installation instructions](https://irkernel.github.io/installation/)
  
- [x] I reproduced the error with the newest versions of `IRkernel`, `IRdisplay`, and `repr`
  
- [x] I included a [minimal reproducible example](https://stackoverflow.com/a/5963610/247482)
- [x] This the right repository: I think the issue is with IRkernel, not
      [IRkernel/repr](https://github.com/IRkernel/repr/issues/new) or a third party repository

**Comment by flying-sheep** [comment]
Good point, this is currently not really supported, since the system relies on `repr::mime2repr``, which is a (immutable) package variable. We could implement extensibility in one of several ways:

- Make `mime2repr` an option (e.g. called `jupyter.mime2repr`)
- Make `repr::mime2repr` mutable by making it a by-reference R6 object
- Add an way to register 3rd party mimetypes, e.g. a function or an option

Aside:

> The first two mean you also need to add a new mime type to the `jupyter.display_mimetypes` option, while the third could handle both.
> 
> ```r
> DISPLAY_TABLE_MIMETYPE <- 'our_custom_mimetype'
> options(jupyter.display_mimetypes=c(getOption('jupyter.display_mimetypes'), DISPLAY_TABLE_MIMETYPE))
> ```

What do you think?

**Comment by sijow** [comment]
I'm also interested in this for a package providing a `text/vnd.typst` representation for ggplot objects. A registration API seems the most natural to me: packages should be able to register MIME renderers without modifying `repr::mime2repr` . Ideally, registration would also add the new MIME type to the display list (with an option to opt out) so users are not required to manually edit `jupyter.display_mimetypes`.
